src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs

```diff
@@ -0,0 +1,375 @@
+
+                    // If the buffer is not empty, then we copy the bytes from the buffer
+                    // to the current chunk being processed.
+                    if (this.buffer != null)
+                    {
+                        ReadOnlyMemory<byte> bytesNotProcessedFromPreviousChunk = this.buffer.AsMemory().Slice(bufferPosition - bytesNotWrittenFromPreviousChunk, bytesNotWrittenFromPreviousChunk);
+
+                        int totalLength = bytesNotProcessedFromPreviousChunk.Length + chunk.Length;
+
+                        // Rent an array to hold bytes from both previous and current chunks.
+                        byte[] combinedArray = ArrayPool<byte>.Shared.Rent(totalLength);
+
+                        // Copy bytes from bytesNotProcessedFromPreviousChunk to the combined array
+                        bytesNotProcessedFromPreviousChunk.Span.CopyTo(combinedArray);
+
+                        // Copy bytes from chunk to the combined array, starting from the end of bytesNotProcessedFromPreviousChunk
+                        chunk.Span.CopyTo(combinedArray.AsSpan(bytesNotProcessedFromPreviousChunk.Length));
+
+                        WriteChunk(combinedArray, totalLength, isFinalBlock);
+
+                        if (combinedArray != null)
+                        {
+                            ArrayPool<byte>.Shared.Return(combinedArray);
+                        }
+                    }
+                    else
+                    {
+                        WriteChunk(chunk.Span, chunk.Length, isFinalBlock);
+                    }
+
+                    // Flush the writer if the buffer threshold is reached
+                    await this.jsonWriter.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+                }
+            }
+
+            /// <summary>
+            /// Writes a chunk of data, encoding it using Base64, and updates the buffer with any unprocessed bytes.
+            /// </summary>
+            /// <param name="chunk">The chunk of data to be encoded and written.</param>
+            /// <param name="isFinalBlock">A boolean indicating whether this is the final block of data.</param>
+            private void WriteChunk(ReadOnlySpan<byte> chunk, int chunkLength, bool isFinalBlock)
+            {
+                // Encode the current chunk using Base64 and write it
+                this.jsonWriter.Base64EncodeAndWriteChunk(chunk.Slice(0, chunkLength), isFinalBlock, out bytesNotWrittenFromPreviousChunk);
+
+                if (bytesNotWrittenFromPreviousChunk > 0)
+                {
+                    if (this.buffer == null)
+                    {
+                        this.buffer = new byte[chunkSize];
+                    }
+
+                    // Update the buffer with unprocessed bytes from the current chunk.
+                    chunk.Slice(chunkLength - bytesNotWrittenFromPreviousChunk).CopyTo(this.buffer.AsSpan(bufferPosition));

```

Comment:
Why append these leftover bytes to the exisitng buffer contents instead of replacing the buffer contents altogether, such that we always write and read from buffer position 0 and only keep track of the length of the contents in the buffer?